### DISCUSSION TITLE
Batch normalization in CNN (DNA chapter)

### DISCUSSION THREAD

**Discussion by imteekay** [deep-learning-for-biology/notebooks#10] [Q&A] [answered]
I'm reading the chapter on DNA and implementing the second version of the CNN model (with batch normalization, dropout, and LR scheduling), and noticed that the book (and the [implementation](https://github.com/deep-learning-for-biology/dlfb/blob/main/src/dlfb/dna/model.py#L74-L75)) adds batch normalization before the gelu function in the first convolutional layer, but adds the same normalization [after gelu](https://github.com/deep-learning-for-biology/dlfb/blob/main/src/dlfb/dna/model.py#L82-L83) in the second conv layer.

Is that correct? If so, what's the intuition on why adding before gelu in the first layer and after in the second? Thanks!

**Comment by cnjr2** [comment, marked as answer]
Good catch, @imteekay - this isn’t an intentional design choice but rather an inconsistency in the implementation. Conventionally, BatchNorm is applied before the activation (Conv → BatchNorm → GELU). The reasoning is that BatchNorm normalizes the pre-activation values (so they have a stable mean and variance), and then the activation operates on those normalized inputs, which generally improves training stability and gradient flow. In practice, this ordering difference will almost certainly have no measurable effect on results, as the network can easily adapt either way.

That said, for clarity and best practices, BatchNorm should come before the activation in both convolutional blocks, and your changes in https://github.com/deep-learning-for-biology/dlfb/pull/9 are spot on. I will bring in your changes in a parallel branch for a future release (as opposed to immediately), since the change will require a rerun of the models (and upload) and lead to slight changes in plots that might confuse readers. In the meantime I have added a comment to the relevant code here https://github.com/deep-learning-for-biology/dlfb/pull/12/commits/598670a84577b06faa02fc2c197bfea918909ffe.

Thank you for your contribution :bow:
